File: scanner.go
package main

import (
	"fmt"
	"strings"
)

const (
	EmptyData = 0
)

var KeywordMap = map[string]TokenType{
	"class":  CLASS,
	"else":   ELSE,
	"false":  FALSE,
	"for":    FOR,
	"fun":    FUNCTION,
	"if":     IF,
	"null":   NULL,
	"print":  PRINT,
	"return": RETURN,
	"super":  SUPER,
	"this":   THIS,
	"true":   TRUE,
	"var":    VAR,
	"while":  WHILE,
	"do":     DO,
}

type Scanner struct {
	source  []rune
	tokens  []*Token
	start   int
	current int
	line    int
}

func NewScanner(source string) *Scanner {
	var tokens []*Token
	return &Scanner{
		source:  []rune(source),
		tokens:  tokens,
		start:   0,
		current: 0,
		line:    1,
	}
}

func (scanner *Scanner) isAtEnd() bool {
	return scanner.current >= len(scanner.source)
}
func (scanner *Scanner) getChar(index int) rune {
	return scanner.source[index]
}
func (scanner *Scanner) peek() rune {
	if scanner.isAtEnd() {
		return EmptyData
	}
	return scanner.getChar(scanner.current)
}
func (scanner *Scanner) peekNext() rune {
	if scanner.current+1 < len(scanner.source) {
		return scanner.getChar(scanner.current + 1)
	}
	return EmptyData
}
func (scanner *Scanner) advance() rune {
	c := scanner.getChar(scanner.current)
	scanner.current++
	return c
}

func (scanner *Scanner) getSubString(start int, end int) string {
	text := string(scanner.source[start:end])
	return text
}

func (scanner *Scanner) addOneToken(tokenType TokenType) {
	text := scanner.getSubString(scanner.start, scanner.current)
	scanner.appendToken(tokenType, text)
}

func (scanner *Scanner) appendToken(tokenType TokenType, text string) {
	scanner.tokens = append(scanner.tokens, &Token{
		tokenType: tokenType,
		lexeme:    text,
		line:      scanner.line,
	})
}

func (scanner *Scanner) addToken(tokenType TokenType) {
	scanner.addOneToken(tokenType)
}
func (scanner *Scanner) match(char rune) bool {
	if scanner.isAtEnd() {
		return false
	}
	if scanner.getChar(scanner.current) != char {
		return false
	}
	scanner.current++
	return true
}

func (scanner *Scanner) isDigit(c rune) bool {
	return c >= '0' && c <= '9'
}

func (scanner *Scanner) isAlpha(c rune) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '\\' || c == '_' || c == '$' || c == '#' || (c >= '\u4e00' && c <= '\u9fa5')
}

func (scanner *Scanner) number() {
	for scanner.isDigit(scanner.peek()) {
		scanner.advance()
	}
	if scanner.peek() == '.' && scanner.isDigit(scanner.peekNext()) {
		scanner.advance()
		for scanner.isDigit(scanner.peek()) {
			scanner.advance()
		}
		scanner.addToken(FLOAT64)
	} else {
		scanner.addToken(INT64)
	}

}

func (scanner *Scanner) string(end rune) {
	for scanner.peek() != end && !scanner.isAtEnd() {
		if scanner.peek() == '\n' {
			scanner.line++
		}
		scanner.advance()
	}
	if scanner.isAtEnd() {
		fmt.Println("unterminated string")
		return
	}
	scanner.advance() // skip "
	text := scanner.getSubString(scanner.start+1, scanner.current-1)
	scanner.appendToken(STRING, text)
}

func (scanner *Scanner) identifier() {
	for scanner.isAlpha(scanner.peek()) {
		scanner.advance()
	}
	text := scanner.getSubString(scanner.start, scanner.current)
	tokenType := IDENTIFIER
	if val, ok := KeywordMap[text]; ok {
		tokenType = val
	}
	scanner.addToken(tokenType)
}

func (scanner *Scanner) scanToken() {
	c := scanner.advance()
	switch c {
	case '(':
		scanner.addToken(LeftParen)
	case ')':
		scanner.addToken(RightParen)
	case '{':
		scanner.addToken(LeftBrace)
	case '}':
		scanner.addToken(RightBrace)
	case '[':
		scanner.addToken(LeftSquare)
	case ']':
		scanner.addToken(RightSquare)
	case ',':
		scanner.addToken(COMMA)
	case '.':
		scanner.addToken(DOT)
	case '-':
		if scanner.match('-') {
			scanner.addToken(MinusMinus)
		} else {
			scanner.addToken(MINUS)
		}
	case '+':
		if scanner.match('+') {
			scanner.addToken(PlusPlus)
		} else {
			scanner.addToken(PLUS)
		}
	case ';':
		scanner.addToken(SEMICOLON)
	case ':':
		scanner.addToken(COLON)
	case '%':
		scanner.addToken(PERCENT)
	case '?':
		scanner.addToken(MARK)
	case '&':
		if scanner.match('&') {
			scanner.addToken(AND)
		} else {
			scanner.addToken(BitAnd)
		}
	case '|':
		if scanner.match('|') {
			scanner.addToken(OR)
		} else {
			scanner.addToken(BitOr)
		}
	case '*':
		scanner.addToken(STAR)
	case '!':
		if scanner.match('=') {
			scanner.addToken(BangEqual)
		} else {
			scanner.addToken(BANG)
		}
	case '=':
		if scanner.match('=') {
			scanner.addToken(EqualEqual)
		} else {
			scanner.addToken(EQUAL)
		}
	case '>':
		if scanner.match('=') {
			scanner.addToken(GreaterEqual)
		} else {
			scanner.addToken(GREATER)
		}
	case '<':
		if scanner.match('=') {
			scanner.addToken(LessEqual)
		} else {
			scanner.addToken(LESS)
		}
	case '/':
		if scanner.match('/') {
			for scanner.peek() != '\n' && !scanner.isAtEnd() {
				scanner.advance()
			}
			text := scanner.getSubString(scanner.start, scanner.current)
			if strings.Contains(text, "expect:") {
				scanner.appendToken(LineComment, text)
			}
		} else if scanner.match('*') {
			for !((scanner.peek() == '*' && scanner.peekNext() == '/') || scanner.isAtEnd()) {
				scanner.advance()
			}
			scanner.advance() // skip *
			scanner.advance() // skip /
		} else {
			scanner.addToken(SLASH)
		}
	case ' ':
	case '\r':
	case '\t':
		break
	case '\n':
		scanner.line++
	case '\'':
	case '"':
		scanner.string(c)
	default:
		if scanner.isDigit(c) {
			scanner.number()
		} else if scanner.isAlpha(c) {
			scanner.identifier()
		} else {
			fmt.Printf("Unexpected character:%c\n", c)
		}
	}
}

func (scanner *Scanner) ScanTokens() []*Token {
	for !scanner.isAtEnd() {
		scanner.start = scanner.current
		scanner.scanToken()
	}
	scanner.appendToken(EOF, "")
	return scanner.tokens
}

File: src/scanner.ts
import Token from './token';
import { TokenType } from './tokenType';
import { KEYWORD_MAP, EMPTY_DATA } from './constant';

class Scanner {
  readonly source: string[];
  readonly tokens: Token[] = [];
  readonly errors: string[] = [];
  private start = 0;
  private current = 0;
  private line = 1;

  constructor(text: string) {
    // unicode split
    this.source = [...text];
  }
  scan = () => {
    while (!this.isAtEnd()) {
      this.start = this.current;
      this.scanToken();
    }
    this.tokens.push(new Token(TokenType.EOF, '', this.line));
    return this.tokens;
  };
  private addError = (line: number, message: string) => {
    const msg = `line: ${line},scanner error : ${message} `;
    throw new Error(msg);
  };
  private isAtEnd() {
    return this.current >= this.source.length;
  }
  private substr(start = this.start, end = this.current): string {
    return this.source.slice(start, end).join('');
  }
  private addOneToken(type: TokenType, value = this.substr()) {
    this.tokens.push(new Token(type, value, this.line));
  }
  private getChar(index: number) {
    return this.source[index] as string;
  }
  private peek() {
    if (this.isAtEnd()) {
      return EMPTY_DATA;
    }
    return this.getChar(this.current);
  }
  private peekNext() {
    if (this.current + 1 < this.source.length) {
      return this.getChar(this.current + 1);
    }
    return EMPTY_DATA;
  }
  private match(expected: string) {
    if (this.isAtEnd()) {
      return false;
    }
    if (this.getChar(this.current) !== expected) {
      return false;
    }

    this.current++;
    return true;
  }
  private advance() {
    return this.getChar(this.current++);
  }
  private scanToken() {
    const c = this.advance();
    switch (c) {
      case '(':
        this.addOneToken(TokenType.LEFT_BRACKET);
        break;
      case ')':
        this.addOneToken(TokenType.RIGHT_BRACKET);
        break;
      case '{':
        this.addOneToken(TokenType.lEFT_BRACE);
        break;
      case '}':
        this.addOneToken(TokenType.RIGHT_BRACE);
        break;
      case '[':
        this.addOneToken(TokenType.LEFT_SQUARE_BRACKET);
        break;
      case ']':
        this.addOneToken(TokenType.RIGHT_SQUARE_BRACKET);
        break;
      case ':':
        this.addOneToken(TokenType.COLON);
        break;
      case ',':
        this.addOneToken(TokenType.COMMA);
        break;
      case '.':
        this.addOneToken(TokenType.DOT);
        break;
      case '-':
        if (this.match('-')) {
          this.addOneToken(TokenType.MINUS_MINUS);
        } else {
          if (this.match('=')) {
            this.addOneToken(TokenType.MINUS_EQUAL);
          } else {
            this.addOneToken(TokenType.MINUS);
          }
        }
        break;
      case '+':
        if (this.match('+')) {
          this.addOneToken(TokenType.PLUS_PLUS);
        } else {
          if (this.match('=')) {
            this.addOneToken(TokenType.PLUS_EQUAL);
          } else {
            this.addOneToken(TokenType.PLUS);
          }
        }
        break;
      case ';':
        this.addOneToken(TokenType.SEMICOLON);
        break;
      case '*':
        if (this.match('=')) {
          this.addOneToken(TokenType.STAR_EQUAL);
        } else if (this.match('*')) {
          if (this.match('=')) {
            this.addOneToken(TokenType.STAR_STAR_EQUAL);
          } else {
            this.addOneToken(TokenType.STAR_STAR);
          }
        } else {
          this.addOneToken(TokenType.STAR);
        }
        break;
      case '%':
        if (this.match('=')) {
          this.addOneToken(TokenType.REMAINDER_EQUAL);
        } else {
          this.addOneToken(TokenType.REMAINDER);
        }
        break;
      case '!':
        if (this.match('=')) {
          if (this.match('=')) {
            this.addOneToken(TokenType.BANG_EQUAL_EQUAL);
          } else {
            this.addOneToken(TokenType.BANG_EQUAL);
          }
        } else {
          this.addOneToken(TokenType.BANG);
        }
        break;
      case '=':
        if (this.match('=')) {
          if (this.match('=')) {
            this.addOneToken(TokenType.EQUAL_EQUAL_EQUAL);
          } else {
            this.addOneToken(TokenType.EQUAL_EQUAL);
          }
        } else {
          this.addOneToken(TokenType.EQUAL);
        }
        break;
      case '?':
        if (this.match('?')) {
          if (this.match('=')) {
            this.addOneToken(TokenType.NULLISH_COALESCING_EQUAL);
          } else {
            this.addOneToken(TokenType.NULLISH_COALESCING);
          }
        } else {
          this.addOneToken(TokenType.QUESTION_MARK);
        }
        break;
      case '>':
        if (this.match('=')) {
          this.addOneToken(TokenType.GREATER_EQUAL);
        } else if (this.match('>')) {
          if (this.match('>')) {
            if (this.match('=')) {
              this.addOneToken(TokenType.UNSIGNED_RIGHT_SHIFT_EQUAL);
            } else {
              this.addOneToken(TokenType.UNSIGNED_RIGHT_SHIFT);
            }
          } else if (this.match('=')) {
            this.addOneToken(TokenType.RIGHT_SHIFT_EQUAL);
          } else {
            this.addOneToken(TokenType.RIGHT_SHIFT);
          }
        } else {
          this.addOneToken(TokenType.GREATER);
        }
        break;
      case '<':
        if (this.match('=')) {
          this.addOneToken(TokenType.LESS_EQUAL);
        } else if (this.match('<')) {
          if (this.match('=')) {
            this.addOneToken(TokenType.LEFT_SHIFT_EQUAL);
          } else {
            this.addOneToken(TokenType.LEFT_SHIFT);
          }
        } else {
          this.addOneToken(TokenType.LESS);
        }
        break;
      case '/':
        // single line comment
        if (this.match('/')) {
          while (this.peek() !== '\n' && !this.isAtEnd()) {
            this.advance();
          }
        } else if (this.match('*')) {
          /* multiple line comments */
          while (
            !(
              (this.peek() === '*' && this.peekNext() === '/') ||
              this.isAtEnd()
            )
          ) {
            this.advance();
            if (this.peek() === '\n') {
              this.line++;
            }
          }
          if (this.peekNext() !== '/') {
            this.addError(this.line, 'multiple line comment end error');
          }
          this.advance(); // skip *
          this.advance(); // skip /
        } else if (this.match('=')) {
          this.addOneToken(TokenType.SLASH_EQUAL);
        } else {
          this.addOneToken(TokenType.SLASH);
        }
        break;
      case '|':
        if (this.match('|')) {
          if (this.match('=')) {
            this.addOneToken(TokenType.OR_EQUAL);
          } else {
            this.addOneToken(TokenType.OR);
          }
        } else if (this.match('=')) {
          this.addOneToken(TokenType.BIT_OR_EQUAL);
        } else {
          this.addOneToken(TokenType.BIT_OR);
        }
        break;
      case '^':
        if (this.match('=')) {
          this.addOneToken(TokenType.BIT_X_OR_EQUAL);
        } else {
          this.addOneToken(TokenType.BIT_X_OR);
        }
        break;
      case '~':
        this.addOneToken(TokenType.BIT_NOT);
        break;
      case '&':
        if (this.match('&')) {
          if (this.match('=')) {
            this.addOneToken(TokenType.AND_EQUAL);
          } else {
            this.addOneToken(TokenType.AND);
          }
        } else if (this.match('=')) {
          this.addOneToken(TokenType.BIT_AND_EQUAL);
        } else {
          this.addOneToken(TokenType.BIT_AND);
        }
        break;
      case ' ':
      case '\r':
      case '\t':
        // whitespace
        break;
      case '\n':
        this.line++;
        break;
      case '"':
        this.string(c);
        break;
      case "'":
        this.string(c);
        break;
      default:
        if (this.isDigit(c)) {
          this.number();
        } else if (this.identifierChar(c)) {
          this.identifier();
        } else {
          this.addError(this.line, `Unexpected character: ${c}`);
        }
        break;
    }
  }
  private string(splitter: string) {
    while (this.peek() !== splitter && !this.isAtEnd()) {
      if (this.peek() === '\n') {
        this.line++;
      }
      this.advance();
    }
    if (this.isAtEnd()) {
      this.addError(this.line, 'Unterminated string');
      return;
    }
    this.advance();
    const value = this.substr(this.start + 1, this.current - 1);
    this.addOneToken(TokenType.STRING, value);
  }
  private number() {
    while (this.isDigit(this.peek())) {
      this.advance();
    }
    if (this.peek() === '.' && this.isDigit(this.peekNext())) {
      this.advance();
      while (this.isDigit(this.peek())) {
        this.advance();
      }
    }
    this.addOneToken(TokenType.NUMBER);
  }
  private identifier() {
    while (this.identifierChar(this.peek())) {
      this.advance();
    }
    const text = this.substr();
    const temp = KEYWORD_MAP.get(text);
    let type: TokenType = TokenType.IDENTIFIER;
    if (temp !== undefined) {
      type = temp;
    }
    this.addOneToken(type);
  }
  private identifierChar(c: string): boolean {
    const temp = `()[]{},.=-*/%!&<>|';":`;
    return !(this.isWhiteSpace(c) || temp.includes(c));
  }
  private isWhiteSpace(c: string) {
    return c === ' ' || c === '\r' || c === '\n' || c === '\t';
  }
  private isDigit(char: string) {
    return char >= '0' && char <= '9';
  }
}
export default Scanner;
